### PR TITLE
Dem force chains

### PR DESCRIPTION
**Pull request by cleodeletre**
# Description of the problem

Need for a post-processing feature to visualize force chains between particles.This code enables the visualization of force chains for each time step with a vtu file. It is noted that the parallelization is not completely done. Ghost particles are not taken into account for the moment. 

# Description of the solution

Modification of parameters_lagrangian(.cc & .h) : Add a force_chains entry variable for users to enable the post-processing feature in the parameter file (.prm)
Creation of force_chains_visualization(.cc & .h) : New class and its methods to calculate the normal force between particles and write vtu files to display force chains.
Modification of dem.cc : Call force_chains_visualization methods to calculate normal forces and write vtu files.

# How Has This Been Tested?

Tested on two dem examples: Small Scale Rotating Drum & Rectangular Hopper

# Future changes

- Parallelize the feature.

# Comments

- source/dem/test_cd_fc.cc should not be in this PR.

### CODE REVIEW

File: include/core/parameters_lagrangian.h [hunk 1]
```diff
@@ -474,6 +474,8 @@ namespace Parameters
     {
       // A bool variable which sets-up the Lagrangian post-processing
       bool Lagrangian_post_processing;
+      // A bool variable which sets-up the force chains visualization
```
Comment (blaisb) [inline include/core/parameters_lagrangian.h:477]:
```suggestion
      /// A bool variable which sets-up the force chains visualization
```

File: include/dem/force_chains_visualization.h [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+
+#include <boost/range/adaptor/map.hpp>
+
+#include <vector>
+
+using namespace dealii;
+using namespace DEM;
+
+#ifndef particles_force_chains_v2_h
+#  define particles_force_chains__v2_h
```
Comment (blaisb) [inline include/dem/force_chains_visualization.h:21]: Fix the name to remove the _v2

File: include/dem/force_chains_visualization.h [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+template <
+  int                                                       dim,
+  Parameters::Lagrangian::ParticleParticleContactForceModel force_model,
+  Parameters::Lagrangian::RollingResistanceMethod rolling_friction_model>
+class ParticlesForceChains
+  : public ParticleParticleContactForce<dim,
+                                        force_model,
+                                        rolling_friction_model>
+{
+public:
```
Comment (blaisb) [inline include/dem/force_chains_visualization.h:34]: You do not know it yet so that's not a big deal, but try to add a @brief and a description of all the [in] arguments. You can look at other files to see how this is done :)

File: include/dem/force_chains_visualization.h [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+
+  void
+  calculate_force_chains(DEMContactManager<dim> &container_manager,
+                         const double            dt);
+
+
+  void
+  write_force_chains(MPI_Comm           mpi_communicator,
+                     const std::string  folder,
+                     const unsigned int iter);
```
Comment (blaisb) [inline include/dem/force_chains_visualization.h:52]: The MPI_Comm should be const

File: include/dem/force_chains_visualization.h [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+{
+public:
+  ParticlesForceChains(const DEMSolverParameters<dim> &dem_parameters);
+
+  virtual ~ParticlesForceChains()
+  {}
+
+  void
+  multi_general_cell(Triangulation<1, 3>         &tria,
+                     const std::vector<Point<3>> &vertices);
```
Comment (blaisb) [inline include/dem/force_chains_visualization.h:42]: Reference to the triangulation could be const :)

File: include/dem/preliminary_force_chain_visualization.h
Comment (blaisb) [inline include/dem/preliminary_force_chain_visualization.h:1]: You can remove this file i believe :)

File: source/core/parameters_lagrangian.cc [hunk 6]
```diff
@@ -1395,6 +1395,11 @@ namespace Parameters
           "false",
           Patterns::Bool(),
           "State whether Lagrangian post-processing should be performed.");
+        prm.declare_entry(
+          "force chains post-processing",
```
Comment (blaisb) [inline source/core/parameters_lagrangian.cc:1399]: no need for the post-processing because this is already in a post-processing section
I would just write force chains

File: source/dem/CMakeLists.txt [hunk 7]
```diff
@@ -10,6 +10,7 @@ add_library(lethe-dem
   find_boundary_cells_information.cc
   find_cell_neighbors.cc
   find_contact_detection_step.cc
+  force_chains_visualization.cc
```
Comment (blaisb) [inline source/dem/CMakeLists.txt:13]: Thanks for putting this in alphabetical order, appreciate it :)

File: source/dem/dem.cc [hunk 8]
```diff
@@ -1057,6 +1058,21 @@ DEMSolver<dim>::write_output_results()
...
+        dim,
+        Parameters::Lagrangian::ParticleParticleContactForceModel::hertz,
+        Parameters::Lagrangian::RollingResistanceMethod::no_resistance>
+        particles_force_chains_object(parameters);
+      particles_force_chains_object.calculate_force_chains(
+        contact_manager, simulation_control->get_time_step());
+
+      particles_force_chains_object.write_force_chains(this->mpi_communicator,
+                                                       folder,
+                                                       iter);
```
Comment (blaisb) [inline source/dem/dem.cc:1074]: Right now this is hardcoded to use only a single collision model. It is important to refactor this to enable all variants of the force model. Feel free to ask questions regarding this if you need help.

File: source/dem/dem.cc [hunk 9]
```diff
@@ -1448,6 +1464,8 @@ DEMSolver<dim>::solve()
           force,
           periodic_offset);
 
+
+
```
Comment (blaisb) [inline source/dem/dem.cc:1468]: Both lines can be removed :)!

File: source/dem/preliminary_force_chain_visualization.cc [hunk 10]
```diff
@@ -0,0 +1,95 @@
+
+#include <core/manifolds.h>
+#include <core/solutions_output.h>
+
+#include <dem/data_containers.h>
```
Comment (blaisb) [inline source/dem/preliminary_force_chain_visualization.cc:5]: Is this file still necessary? You are not using this code, in that case, I would remove it :)

File: source/dem/test_cd_fc.cc [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+  triangulation.refine_global(4);
+
+  std::ofstream out("hyper_rectangle.svg");
+  GridOut       grid_out;
+  grid_out.write_svg(triangulation, out);
+  std::cout << "Grid written to hyper_rectangle.svg" << std::endl;
+}
+
+void
+generalcell_grid()
```
Comment (blaisb) [inline source/dem/test_cd_fc.cc:47]:
```suggestion
general_cell_grid()
```

File: source/dem/test_cd_fc.cc [hunk 12]
```diff
@@ -0,0 +1,66 @@
...
+{
+  Triangulation<1>            triangulation;
+  const std::vector<Point<1>> vertice<>;
```
Comment (blaisb) [inline source/dem/test_cd_fc.cc:50]: why do you have <> at the end of vertice?
Should also be vertices

File: source/dem/force_chains_visualization.cc [hunk 13]
```diff
@@ -0,0 +1,511 @@
...
+  Parameters::Lagrangian::RollingResistanceMethod rolling_friction_model>
+ParticlesForceChains<dim, contact_model, rolling_friction_model>::
+  ParticlesForceChains(const DEMSolverParameters<dim> &dem_parameters_in)
+  : ParticleParticleContactForce<dim, contact_model, rolling_friction_model>(
+      dem_parameters_in)
+  , dem_parameters(dem_parameters_in)
+{
+  force_normal.push_back(0);
+  vertices.push_back(Point<3>(0, 0, 0));
+  vertices.push_back(Point<3>(0, 0, 0));
```
Comment (blaisb) [inline source/dem/force_chains_visualization.cc:50]: it's allright to initialize to make sure it's never empty, but is this 100% sure necessary?

File: source/dem/force_chains_visualization.cc [hunk 14]
```diff
@@ -0,0 +1,511 @@
...
+
+
+
+template <
+  int                                                       dim,
+  Parameters::Lagrangian::ParticleParticleContactForceModel contact_model,
+  Parameters::Lagrangian::RollingResistanceMethod rolling_friction_model>
+void
+ParticlesForceChains<dim, contact_model, rolling_friction_model>::
+  multi_general_cell(Triangulation<1, 3>         &tria,
```
Comment (blaisb) [inline source/dem/force_chains_visualization.cc:61]: Forget by comment, this cannot be const.

File: source/dem/force_chains_visualization.cc [hunk 15]
```diff
@@ -0,0 +1,511 @@
...
+{
+  const unsigned int       n_cells = vertices.size() / 2;
+  std::vector<CellData<1>> cells(n_cells, CellData<1>());
+  for (unsigned int i = 0; i < n_cells; ++i)
+    {
+      for (unsigned int j = 0; j < 2; ++j)
+        {
+          cells[i].vertices[j] = 2 * i + j;
+          cells[i].material_id = 0;
+        }
```
Comment (blaisb) [inline source/dem/force_chains_visualization.cc:72]: since this loop is always 2, do we really need it? You could replace it with just the 3 lines.

File: source/dem/force_chains_visualization.cc [hunk 16]
```diff
@@ -0,0 +1,511 @@
...
+
+  auto &local_adjacent_particles = container_manager.local_adjacent_particles;
+  // auto &ghost_adjacent_particles =
+  // container_manager.ghost_adjacent_particles; auto
+  // &local_periodic_adjacent_particles =
+  //  container_manager.local_periodic_adjacent_particles;
+  // auto &ghost_periodic_adjacent_particles =
+  //  container_manager.ghost_periodic_adjacent_particles;
+  // auto &ghost_local_periodic_adjacent_particles =
+  //  container_manager.ghost_local_periodic_adjacent_particles;
```
Comment (blaisb) [inline source/dem/force_chains_visualization.cc:100]: If this is not necessary, you can remove those :)

**Review by blaisb — CHANGES_REQUESTED** [review]
First batch of comments, will continue tomorrow :)!

File: include/dem/force_chains_visualization.h [hunk 17]
```diff
@@ -0,0 +1,61 @@
...
+
+  void
+  write_force_chains(MPI_Comm           mpi_communicator,
+                     const std::string  folder,
+                     const unsigned int iter);
+
+private:
+  std::vector<double>             force_normal;
+  std::vector<Point<3>>           vertices;
+  const DEMSolverParameters<dim> &dem_parameters;
```
Comment (OGaboriault) [inline include/dem/force_chains_visualization.h:57]: Add a small description of every attribute in the class. Doesn't need to be long.

Also, do you really need to have `dem_parameters` as an attribute of the class? I think the `ParticleParticleContactForce `object could be an attribute build in the constructor (but maybe I'm wrong)

File: source/dem/force_chains_visualization.cc [hunk 18]
```diff
@@ -0,0 +1,511 @@
...
+#include <deal.II/numerics/data_out.h>
+
+#include <cmath>
+#include <fstream>
+#include <iostream>
+
+using namespace DEM;
+
+
+
```
Comment (OGaboriault) [inline source/dem/force_chains_visualization.cc:37]:
```suggestion
```

Comment (OGaboriault) [reply source/dem/force_chains_visualization.cc:37]: Remove extra lines

File: source/dem/force_chains_visualization.cc [hunk 19]
```diff
@@ -0,0 +1,511 @@
...
+  : ParticleParticleContactForce<dim, contact_model, rolling_friction_model>(
+      dem_parameters_in)
+  , dem_parameters(dem_parameters_in)
+{
+  force_normal.push_back(0);
+  vertices.push_back(Point<3>(0, 0, 0));
+  vertices.push_back(Point<3>(0, 0, 0));
+}
+
+
```
Comment (OGaboriault) [inline source/dem/force_chains_visualization.cc:53]: Remove extra lines

File: source/dem/force_chains_visualization.cc (hunk 16 above)
Comment (OGaboriault) [reply source/dem/force_chains_visualization.cc:100]: I told her to keep those since she soon be working on the ghost cells... my fault

File: source/dem/force_chains_visualization.cc [hunk 20]
```diff
@@ -0,0 +1,511 @@
...
+          Point<3> particle_one_location = [&] {
+            if constexpr (dim == 3)
+              {
+                return particle_one->get_location();
+              }
+            if constexpr (dim == 2)
+              {
+                return (point_nd_to_3d(particle_one->get_location()));
+              }
+          }();
```
Comment (OGaboriault) [inline source/dem/force_chains_visualization.cc:141]: Nice uses of the lambda function, but I don't think it is necessary. You can just use point_nd_to_3d even when dim == 3

File: source/dem/force_chains_visualization.cc [hunk 21]
```diff
@@ -0,0 +1,511 @@
...
+              Point<3> particle_two_location = [&] {
+                if constexpr (dim == 3)
+                  {
+                    return particle_two->get_location();
+                  }
+                if constexpr (dim == 2)
+                  {
+                    return (point_nd_to_3d(particle_two->get_location()));
+                  }
+              }();
```
Comment (OGaboriault) [inline source/dem/force_chains_visualization.cc:161]: Same here

File: source/dem/test_cd_fc.cc [hunk 22]
```diff
@@ -0,0 +1,66 @@
...
+#include <deal.II/grid/tria.h>
+
+#include <cmath>
+#include <fstream>
+#include <iostream>
+
+
+using namespace dealii;
+
+
```
Comment (OGaboriault) [inline source/dem/test_cd_fc.cc:12]: extra lines

**Review by OGaboriault — CHANGES_REQUESTED** [review]
Good work for your first PR!

File: include/dem/force_chains_visualization.h (hunk 5 above)
Comment (blaisb) [reply include/dem/force_chains_visualization.h:42]: actually it cannot be const, that's my bad.

File: source/dem/force_chains_visualization.cc (hunk 16 above)
Comment (blaisb) [reply source/dem/force_chains_visualization.cc:100]: No it's allright, we can keep it. Just add a comment asto why we are keeping this for now.

File: source/dem/force_chains_visualization.cc (hunk 20 above)
Comment (blaisb) [reply source/dem/force_chains_visualization.cc:141]: no let's leave it this way. We do this everywhere in the contact code, let's not change it please.

File: source/dem/force_chains_visualization.cc (hunk 21 above)
Comment (blaisb) [reply source/dem/force_chains_visualization.cc:161]: Please don't leave it like it is now, this is the right way.

File: source/dem/force_chains_visualization.cc [hunk 23]
```diff
@@ -0,0 +1,500 @@
...
+
+              if (normal_overlap > 0.0)
+                {
+                  // This means that the adjacent particles are in contact
+
+
+
+                  // Since the normal overlap is already calculated, we update
+                  // this element of the container here. The rest of information
+                  // are updated using the following function
```
Comment (blaisb) [inline source/dem/force_chains_visualization.cc:167]: there seems to be a lot of empty lines here for no reason.

File: source/dem/force_chains_visualization.cc [hunk 24]
```diff
@@ -0,0 +1,500 @@
...
+                  vertices.push_back(particle_two_location);
+                  force_normal.push_back(sqrt(normal_force.norm()));
+                }
+
+              else
+                {
+                  // if the adjacent pair is not in contact anymore, only the
+                  // tangential overlap is set to zero
+                  contact_info.tangential_overlap.clear();
+                }
```
Comment (blaisb) [inline source/dem/force_chains_visualization.cc:304]: this is not important in this case because you don't have a tangential overlap .

**Review by blaisb — APPROVED** [review]
just 2-3 last comments. Nothing big. Could you add an application test even if it won't deplay anything? Just modify an existing test and add a force chain output. This will enable us to ensure that at least the feature does not scratch. @OGaboriault can give you a hand with this. It can be something very small / minor.

**Comment by blaisb** [comment]
Will merge after @OGaboriault accepts

File: include/core/parameters_lagrangian.h [hunk 25]
```diff
@@ -474,6 +474,8 @@ namespace Parameters
     {
       // A bool variable which sets-up the Lagrangian post-processing
       bool Lagrangian_post_processing;
+      /// A bool variable which sets-up the force chains visualization
```
Comment (OGaboriault) [inline include/core/parameters_lagrangian.h:477]:
```suggestion
      // A bool variable which sets-up the force chains visualization
```

File: include/dem/force_chains_visualization.h [hunk 26]
```diff
@@ -0,0 +1,132 @@
...
+   * vectors.
+   *
+   * @param container_manager The container manager object that contains
+   * containers to modify of contact pair periodic candidates with other
+   * containers with periodic neighbors lists
+   * @param dt DEM time step
+   */
+  virtual void
+  calculate_force_chains(DEMContactManager<dim> &container_manager,
+                         const double            dt) = 0;
```
Comment (OGaboriault) [inline include/dem/force_chains_visualization.h:47]: I think this function doesn't need the `dt` as a parameter. 

I would delete this parameter. You use `dt` at line 165 for `update_contact_information` function which is used to update the tangential overlap. The tangential forces are not being used for the forces chains.

 I would just replace the `dt` by `0.` when calling `update_contact_information`.

**Review by OGaboriault — CHANGES_REQUESTED** [review]
Tiny changes

**Review by OGaboriault — APPROVED** [review]

File: include/dem/force_chains_visualization.h [hunk 27]
```diff
@@ -0,0 +1,130 @@
...
+ * ParticlesForceChains class which is templated by the contact model
+ * type.
+ */
+template <int dim>
+class ParticlesForceChainsBase
+{
+public:
+  /**
+   * @brief Use a ParticleParticleContactForce object to calculate normal forces between
+   * all touching particles. Stock normal forces and particles position in
```
Comment (blaisb) [inline include/dem/force_chains_visualization.h:37]:
```suggestion
   * all touching particles. Store normal forces and particles position in
```

File: include/dem/force_chains_visualization.h [hunk 28]
```diff
@@ -0,0 +1,129 @@
...
+ * @brief Class that carries out the calculation of
+ * particle-particle contact force and the visualization of force chains
+ * by writing vtu files. Instead of using a inheritance hiearchy to distinguish
+ * between the contact model, the class is templated with the type of force
+ * model and rolling friction model. Consequently, the code for each combination
+ * of force model is generated at compile time.
+ *
+ * @tparam dim The dimension of the problem
+ * @tparam force_model The particle-particle contact force model
+ * @tparam rolling_friction_model The rolling resistance model used
```
Comment (blaisb) [inline include/dem/force_chains_visualization.h:69]:
```suggestion
 * @tparam rolling_friction_model The rolling resistance model
```

File: include/dem/force_chains_visualization.h [hunk 29]
```diff
@@ -0,0 +1,129 @@
...
+  void
+  write_force_chains(const MPI_Comm     mpi_communicator,
+                     const std::string  folder,
+                     const unsigned int iter) override;
+
+private:
+  // vector of normal forces between each touching particles.
+  std::vector<double> force_normal;
+  // vector of positions of touching particles.
+  std::vector<Point<3>> vertices;
```
Comment (blaisb) [inline include/dem/force_chains_visualization.h:125]:
```suggestion
  /// vector of normal forces between each touching particles.
  std::vector<double> force_normal;
  /// vector of positions of touching particles.
  std::vector<Point<3>> vertices;
```

**Review by blaisb — APPROVED** [review]
I fixed some last minute typos, will merge after CI
